pkg/hosts/websocket.go

```diff
@@ -0,0 +1,307 @@
+package hosts
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os/exec"
+
+	"github.com/cnrancher/autok3s/pkg/types"
+
+	"github.com/gorilla/websocket"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	SSHKind    = "ssh"
+	DockerKind = "docker"
+	PtyKind    = "pty"
+)
+
+type WebSocketDialer struct {
+	kind         string
+	sshDialer    *SSHDialer
+	dockerDialer *DockerDialer
+	ptyDialer    *PtyDialer
+
+	conn    *websocket.Conn
+	session *ssh.Session
+	reader  *TerminalReader
+
+	err error
+}
+
+func NewWebSocketDialer(kind string, n *types.Node, conn *websocket.Conn, cmd *exec.Cmd) (*WebSocketDialer, error) {
+	d := &WebSocketDialer{
+		kind: kind,
+		conn: conn,
+	}
+
+	r := NewTerminalReader(d.conn)
+	r.SetResizeFunction(d.ChangeWindowSize)
+	w := NewBinaryWriter(d.conn)
+
+	d.reader = r
+
+	switch kind {
+	case SSHKind:
```

Comment:
I think it's better to define interface function for the tree kind of dialer to escape from switch case block in each function.

Reply:
Will optimize when I have time, currently not optimize.